referenceframe/urdf.go

```diff
@@ -0,0 +1,299 @@
+		Upper   float64  `xml:"upper,attr"` // translation limits are in meters, revolute limits are in radians
+	} `xml:"limit"`
+}
+
+// ParseURDFFile will read a given file and parse the contained URDF XML data into an equivalent ModelConfig struct.
+func ParseURDFFile(filename, modelName string) (Model, error) {
+	//nolint:gosec
+	xmlData, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to read URDF file")
+	}
+	return ConvertURDFToConfig(xmlData, modelName)
+}
+
+// ConvertURDFToConfig will transfer the given URDF XML data into an equivalent ModelConfig. Direct unmarshaling in the
+// same fashion as ModelJSON is not possible, as URDF data will need to be evaluated to accommodate differences
+// between the two kinematics encoding schemes.
+func ConvertURDFToConfig(xmlData []byte, modelName string) (Model, error) {
+	// empty data probably means that the read URDF has no actionable information
+	if len(xmlData) == 0 {
+		return nil, ErrNoModelInformation
+	}
+
+	mc := &ModelConfig{}
+	urdf := &URDFConfig{}
+	err := xml.Unmarshal(xmlData, urdf)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to convert URDF data to equivalent URDFConfig struct")
+	}
+
+	if modelName == "" {
+		modelName = urdf.Name
+	}
+
+	parentMap := map[string]string{}
+
+	// Migrate URDF elements into an equivalent ModelConfig representation
+	mc.Name = modelName
+	mc.KinParamType = "SVA"
+
+	// Handle joints
+	for _, jointElem := range urdf.Joints {
+		// Checking for reserved names in this or adjacent elements
+		if jointElem.Name == World {
+			return nil, errors.New("Joints with the name 'world' are not supported by config parsers")
+		}
+
+		// Relationship tracking
+		parentMap[jointElem.Name] = jointElem.Parent.Link
+		parentMap[jointElem.Child.Link] = jointElem.Name
+
+		// Set up the child link mentioned in this joint; fill out the details in the link parsing section later
+		childLink := JsonLink{ID: jointElem.Child.Link, Parent: jointElem.Name}
+
+		switch jointElem.Type {
+		case "continuous", "revolute", "prismatic":
+			// Parse important details about each joint, including axes and limits
+			jointAxes := convStringAttrToFloats(jointElem.Axis.XYZ)
+			thisJoint := JsonJoint{
+				ID:     jointElem.Name,
+				Type:   jointElem.Type,
+				Parent: jointElem.Parent.Link,
+				Axis:   spatial.AxisConfig{jointAxes[0], jointAxes[1], jointAxes[2]},
+			}
+
+			// Slightly different limits handling for continuous, revolute, and prismatic joints
+			if jointElem.Type == "continuous" {
+				thisJoint.Type = "revolute" // Currently, we treate a continuous joint as a special case of the revolute joint
+				thisJoint.Min, thisJoint.Max = utils.RadToDeg(-math.Pi), utils.RadToDeg(math.Pi)
+			} else if jointElem.Type == "prismatic" {
+				thisJoint.Min, thisJoint.Max = jointElem.Limit.Lower*1000, jointElem.Limit.Upper*1000 // from meters to mm
+			} else {
```

Comment:
Let's do an explicit check for "revolute", and return an error if it's none of these three.

I know the switch statement above covers that, but it could change.